core/server/common/src/main/java/alluxio/master/transport/GrpcMessagingContext.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.transport;
+
+import com.google.common.base.Preconditions;
+import io.atomix.catalyst.serializer.Serializer;
+import org.apache.http.concurrent.Cancellable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * The context for Grpc messaging single thread.
+ * This context uses a {@link ScheduledExecutorService} to schedule events on the context thread.
+ */
+public class GrpcMessagingContext {
```

Comment:
Is this class needed? Will there be any problem running everything without this context?

Reply:
This class mainly store two things, one executor, and one serializer. 
In all the places, the current code makes sure that we read the response and write request to the same serializer.
all the requests are handled by the same executor, all the listeners acceptances are handled by the same executor.
It's more for the correctness and reduces the potential race conditions.

Reply:
I took a deeper look into the code, the backup leader may create more than one bi-di stream (in case need to reconnect?) with the client, but all the streams shared the same context (same executor and serializer).

Reply:
I see. We can leave it as it for now. 

For serializer we will probably replace it with protobuf built-in serialization. The gRPC also provide some guarantee in message order, so this class may eventually become unnecessary.